connector/spanmetricsconnector/connector_test.go

```diff
@@ -1259,3 +1265,70 @@ func TestConnector_initHistogramMetrics(t *testing.T) {
 		})
 	}
 }
+
+func TestSpanMetrics_Events(t *testing.T) {
+	tests := []struct {
+		name                   string
+		eventsConfig           Event
+		shouldEventMetricExist bool
+	}{
+		{
+			name:                   "events disabled",
+			eventsConfig:           Event{Enabled: false, Dimensions: []Dimension{{Name: "exception.type", Default: stringp("NullPointerException")}}},
+			shouldEventMetricExist: false,
+		},
+		{
+			name:                   "events without dimensions",
+			eventsConfig:           Event{Enabled: true, Dimensions: []Dimension{}},
+			shouldEventMetricExist: false,
+		},
+		{
+			name:                   "events with dimensions",
+			eventsConfig:           Event{Enabled: true, Dimensions: []Dimension{{Name: "exception.type", Default: stringp("NullPointerException")}}},
+			shouldEventMetricExist: true,
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			// Prepare
+			mcon := &consumertest.MetricsSink{}
+
+			mockClock := clock.NewMock(time.Now())
+			ticker := mockClock.NewTicker(time.Nanosecond)
+
+			p := newConnectorImp(t, mcon, stringp("defaultNullValue"), explicitHistogramsConfig, disabledExemplarsConfig, tt.eventsConfig, cumulative, zaptest.NewLogger(t), ticker)
+
+			ctx := metadata.NewIncomingContext(context.Background(), nil)
+			err := p.Start(ctx, componenttest.NewNopHost())
+			defer func() { sdErr := p.Shutdown(ctx); require.NoError(t, sdErr) }()
+			require.NoError(t, err)
+
+			err = p.ConsumeTraces(ctx, buildSampleTrace())
+			require.NoError(t, err)
+			metrics := p.buildMetrics()
+			for i := 0; i < metrics.ResourceMetrics().Len(); i++ {
+				rm := metrics.ResourceMetrics().At(i)
+				ism := rm.ScopeMetrics()
+				for ilmC := 0; ilmC < ism.Len(); ilmC++ {
+					m := ism.At(ilmC).Metrics()
+					if tt.shouldEventMetricExist {
+						assert.Equal(t, m.Len(), 3)
+						for mC := 0; mC < m.Len(); mC++ {
+							metric := m.At(mC)
+							if metric.Name() == "events" {
+								assert.Equal(t, metric.Type(), pmetric.MetricTypeSum)
+								for idp := 0; idp < metric.Sum().DataPoints().Len(); idp++ {
+									attrs := metric.Sum().DataPoints().At(idp).Attributes()
+									assert.Contains(t, attrs.AsRaw(), exceptionTypeAttrName)
+								}
+							}
+						}
+					} else {
+						assert.Equal(t, m.Len(), 2)
+					}
+				}
+			}
+
```

Comment:
- We can reduce the nesting to reduce cognitive load on readers.
- `assert.Equals` should have the expectation before the actual value: it's borrowed from the junit `assertEquals` convention I think.

Putting that together, here's a suggested example:
```suggestion
			for i := 0; i < metrics.ResourceMetrics().Len(); i++ {
				rm := metrics.ResourceMetrics().At(i)
				ism := rm.ScopeMetrics()
				for ilmC := 0; ilmC < ism.Len(); ilmC++ {
					m := ism.At(ilmC).Metrics()

					if !tt.shouldEventMetricExist {
						assert.Equal(t, 2, m.Len())
						continue
					}

					assert.Equal(t, 3, m.Len())
					for mC := 0; mC < m.Len(); mC++ {
						metric := m.At(mC)
						if metric.Name() != "events" {
							continue
						}
						assert.Equal(t, pmetric.MetricTypeSum, metric.Type())
						for idp := 0; idp < metric.Sum().DataPoints().Len(); idp++ {
							attrs := metric.Sum().DataPoints().At(idp).Attributes()
							assert.Contains(t, attrs.AsRaw(), exceptionTypeAttrName)
						}
					}
				}
			}
```